libpages/server.go

```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+	fs, err := s.getFS(r.Context(), root)
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+	http.FileServer(fs).ServeHTTP(w, r)
+}
+
+// ErrDomainNotAllowedInWhitelist is returned when the server is configured
+// with a domain whitelist, and we receive a HTTP request that was sent to a
+// domain that's not in the whitelist.
+type ErrDomainNotAllowedInWhitelist struct{}
+
+// Error implements the error interface.
+func (ErrDomainNotAllowedInWhitelist) Error() string {
+	return "a whitelist is configured and the given domain is not in the list"
+}
+
+func (s *Server) allowedByDomainWhiteList(domain string) bool {
+	s.whiteListOnce.Do(func() {
+		if len(s.config.DomainWhitelist) > 0 {
+			s.whiteList = make(map[string]bool)
+			for _, d := range s.config.DomainWhitelist {
+				s.whiteList[d] = true
+			}
+		}
+	})
+	if s.whiteList != nil {
+		return s.whiteList[domain]
+	}
+
+	// No whitelist; allow everything!
+	return true
+}
+
+// allowConnection is used as a HostPolicy in autocert package.
+func (s *Server) allowConnectionTo(ctx context.Context, host string) error {
+	if !s.allowedByDomainWhiteList(host) {
+		return ErrDomainNotAllowedInWhitelist{}
+	}
+
+	// DoS protection: look up kbp TXT record before attempting ACME cert
+	// issuance, and only allow those that have DNS records configured. This is
+	// in case someone keeps sending us TLS handshakes with random SNIs,
+	// causing us to be rate-limited by the ACME server.
+	//
+	// TODO: cache the parsed root somewhere so we don't end up doing it twice
+	// for each connection.
+	_, err := LoadRootFromDNS(s.config.Logger, host)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const (
+	gracefulShutdownTimeout = 10 * time.Second
+)
+
+// ListenAndServe listens on 443 and 80 ports of all addresses, and serve
+// Keybase Pages based on config and kbfsConfig. HTTPs setup is handled with
+// ACME.
+func ListenAndServe(ctx context.Context,
+	config ServerConfig, kbfsConfig libkbfs.Config) (err error) {
+	fsCache, err := lru.New(fsCacheSize)
+	if err != nil {
+		return err
+	}
+	server := &Server{
+		config:     config,
+		kbfsConfig: kbfsConfig,
+		fsCache:    fsCache,
+	}
+
+	if config.AutoDirectHTTP {
+		// TODO
+	}
+
+	manager := autocert.Manager{
+		Prompt:     autocert.AcceptTOS,
+		HostPolicy: server.allowConnectionTo,
+	}
+
+	if config.UseDiskCacheForDev {
+		manager.Cache = autocert.DirCache("./kbp-cert-cache-dev")
+	}
+
+	if config.UseStaging {
+		acmeKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return err
+		}
+		manager.Client = &acme.Client{
+			DirectoryURL: "https://acme-staging.api.letsencrypt.org/directory",
+			Key:          acmeKey,
+		}
+	}
+
+	httpServer := http.Server{
+		Handler: server,
+	}
+
+	go func() {
+		<-ctx.Done()
+		shutdownCtx, cancel := context.WithTimeout(
+			context.Background(), gracefulShutdownTimeout)
```

Comment:
I don't really understand how this graceful shutdown works.  How will the caller (in this case, the `main` function) know that shutdown is complete?  Wouldn't we need to return a channel that gets closed or something?

Reply:
The `ListenAndServe` function blocks until everything is finished. If `ctx` is canceled in `kbp` package, the goroutine (main) will block on this function until everything is shutdown, or `gracefulShutdownTimeout` has passed, whichever happens first.

Reply:
I guess you mean `httpServer.Serve()` will block until everything is finished?  There's no scenario where it returns early without `ctx` being canceled, but where a `Shutdown` call is still needed?

Reply:
Yes, `httpServer.Serve()` is where it blocks.

> There's no scenario where it returns early without ctx being canceled, but where a Shutdown call is still needed?

Hmm perhaps we should cancel the ctx in a defered call in `ListenAndServe` function too?